lang/semgrep-grammars/src/semgrep-dockerfile/grammar.js

```diff
@@ -35,52 +35,147 @@ module.exports = grammar(base_grammar, {
       $.semgrep_metavariable
     ),
 
-    // TODO: support metavariables and ellipses in a bunch of places
+    // redefinition
+    arg_instruction: ($, previous) => seq(
+      alias(/[aA][rR][gG]/, "ARG"),
+      field(
+        "name",
+        alias(
+          choice(
+            $.semgrep_metavariable,
+            /[a-zA-Z0-9_]+/
+          ),
+          $.unquoted_string
+        )
+      ),
+      optional(
+        seq(
+          token.immediate("="),
+          field(
+            "default",
+            choice(
+              $.double_quoted_string,
+              $.unquoted_string
+            )
+          )
+        )
+      )
+    ),
+
+    // override
+    label_pair: ($, previous) => choice(
+      $.semgrep_ellipsis,
+      seq(
+        field(
+          "key",
+          choice(
+            $.semgrep_metavariable,
+            alias(/[-a-zA-Z0-9\._]+/, $.unquoted_string)
+          )
+        ),
+        token.immediate("="),
+        field("value", choice($.double_quoted_string, $.unquoted_string))
+      )
+    ),
+
+    // TODO: find a way to support metavariables as env keys.
+    env_pair: ($, previous) => choice(
+      $.semgrep_ellipsis,
+      previous
+    ),
+
+    expose_port: ($, previous) => choice(
+      $.semgrep_ellipsis,
+      previous
+    ),
+
+    // override
+    healthcheck_instruction: ($) =>
+      seq(
+        alias(/[hH][eE][aA][lL][tT][hH][cC][hH][eE][cC][kK]/, "HEALTHCHECK"),
+        choice(
+          $.semgrep_metavariable,
+          "NONE",
+          seq(repeat($.param), $.cmd_instruction)
+        )
+      ),
+
+    /*
+      Metavariable syntax vs. ARG expansions:
+
+      - in a semgrep pattern, $FOO is always a metavariable.
+      - in a semgrep pattern, ${FOO} is always an arg expansion.
+      - in a semgrep pattern, $FOO in a position where a metavariable
+        is not allowed results in an error.
+
+     TODO: support metavariables and ellipses in a bunch of places
+    */
     semgrep_metavariable: $ => /\$[A-Z_][A-Z_0-9]*/,
     semgrep_ellipsis: $ => '...',
 
     /*
       TODO: more syntax ellipsis (e) or metavariables (mv):
 
-      e, mv:
-      LABEL multi.label1="value1" multi.label2="value2" other="value3"
+      mv: done
```

Comment:
Are those mv and e defined somewhere? After a few seconds I guessed it meant metavariables and ellipsis
but could be nice to write it somewhere.

Reply:
yes, it's on the line above. :wink:

Reply:
I'll remove these comments when I'm done with the translation to the semgrep AST.

Reply:
oops ...